examples/notebooks/beam-ml/rag_usecase/opensearch_connector.py

```diff
@@ -0,0 +1,372 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+from __future__ import absolute_import
+
+import apache_beam as beam
+
+from apache_beam.transforms import DoFn
+from apache_beam.transforms import PTransform
+from apache_beam.transforms import Reshuffle
+
+from typing import Optional
+from opensearchpy import OpenSearch
+
+import os
+from dotenv import load_dotenv
+
+load_dotenv()
+
+# Set the logging level to reduce verbose information
+import logging
+
+logging.root.setLevel(logging.INFO)
+logger = logging.getLogger(__name__)
+
+__all__ = ['InsertDocInOpenSearch', 'InsertEmbeddingInOpenSearch']
+
+"""This module implements IO classes to read document in Opensearch.
+
+
+Insert Doc in OpenSearch:
+-----------------
+:class:`InsertDocInOpenSearch` is a ``PTransform`` that writes key and values to a
+configured sink, and the write is conducted through a Opensearch pipeline.
+
+The ptransform works by getting the first and second elements from the input,
+this means that inputs like `[k,v]` or `(k,v)` are valid.
+
+Example usage::
+
+  pipeline | InsertDocInOpenSearch(host='localhost',
+                          port=6379,
+                          username='admin',
+                          password='admin'
+                          batch_size=100)
+
+
+No backward compatibility guarantees. Everything in this module is experimental.
+"""
+
+
+class InsertDocInOpenSearch(PTransform):
+    """InsertDocInOpensearch is a ``PTransform`` that writes a ``PCollection`` of
+    key, value tuple or 2-element array into a Opensearch server.
+    """
+
+    def __init__(self,
+                 host: str,
+                 port: int,
+                 username: Optional[str] = os.getenv("OPENSEARCH_USERNAME"),
+                 password: Optional[str] = os.getenv("OPENSEARCH_PASSWORD"),
```

Comment:
You lose some nuance here in terms of typing. Optional[str] is syntactic sugar for `str | None`, setting a default can indicate that you don't necessarily want `None` here (and for this use-case may not be helpful.) I'd recommend leaving the `Optional[str]` typing and moving the `os.getenv` blocks into the function itself as input validation. If a username/password isn't provided, check the environmental variable. If there isn't a value in the environment, react accordingly.

Reply:
Good point, fixing in all places.